apollo-router/src/spec/query.rs

```diff
@@ -829,6 +828,17 @@ impl Query {
                 }
             }
         }
+        // Detect if root __typename is asked in the original query (the qp doesn't put root __typename in subselections)
+        // cf https://github.com/apollographql/router/issues/1677
+        let operation_kind_if_root_typename = self.operations.get(0).and_then(|op| {
```

Comment:
could this be moved (again, sorry :pray:) under the match branch on `Selection::Field`?
The current operation is already selected at line 670 (and it might not be the first one in the list)

Reply:
I think it could be fixed for the `InlineFragment` and `FragmentSpread` as well, by going through the selection set before calling `apply_selection_set`. That would not work with nested fragments though